posthog/models/historical_version.py

```diff
@@ -0,0 +1,44 @@
+from django.db import models
+from django.utils import timezone
+
+
+class HistoricalVersion(models.Model):
+    """
+    We don't _only_ store foreign references cos the referenced model might get deleted.
+    The history log should be relatively immutable
+
+    Everything in the log must have either a team id or an organization id
+    """
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(
+                fields=["organization_id", "team_id", "name", "versioned_at"], name="unique_version"
+            ),
+            models.CheckConstraint(
+                check=models.Q(team_id__isnull=False) | models.Q(organization_id__isnull=False),
+                name="must_have_team_or_organization_id",
+            ),
+        ]
+
+    # JSON of the historical item
+    state = models.JSONField(null=False)
+    # e.g. feature_flags
+    name = models.fields.TextField(null=False)
+    # TODO will this only be create, update, or delete
+    action = models.fields.TextField(null=False)
+
+    # the id of the item being versioned
+    item_id = models.fields.PositiveIntegerField(null=False)
```

Comment:
This is problematic - we have some objects id-d by numbers, others by UUID, some by short ids. This solution does not work.

Reply:
If I'm reading that correctly the generic foreign key object_id column needs to be able to hold a value of any type it might store. I think in this case that means a char field large enough to hold a UUID

I'd be tempted to avoid the foreign key and have a char field in that case.

We're not going to be reading from this table super frequently so 🤷‍♀️ and we don't query along the foreign key reference since the log has to hold the state at the time it was written not at the time it is read

I'll make that change while this is fresh in my mind, and then can pick up next week if folk disagree